cas-server-webapp-actions/src/main/java/org/jasig/cas/web/flow/GenerateServiceTicketAction.java

```diff
@@ -39,19 +41,37 @@
     private CentralAuthenticationService centralAuthenticationService;
 
     @NotNull
-    @Autowired(required=false)
+    @Autowired
     @Qualifier("defaultAuthenticationSystemSupport")
     private AuthenticationSystemSupport authenticationSystemSupport = new DefaultAuthenticationSystemSupport();
 
+    @Autowired
+    @Qualifier("defaultAuthenticationSupport")
+    private TicketRegistrySupport ticketRegistrySupport;
+
     @Override
     protected Event doExecute(final RequestContext context) {
         final Service service = WebUtils.getService(context);
         final String ticketGrantingTicket = WebUtils.getTicketGrantingTicketId(context);
 
         try {
-            final Credential credential = WebUtils.getCredential(context);
-            final AuthenticationResult authenticationResult =
-                    this.authenticationSystemSupport.handleAndFinalizeSingleAuthenticationTransaction(service, credential);
+            /**
+             * In the initial primary authentication flow, credentials are cached and available.
+             * Since they are authenticated as part of submission first, there is no need to doubly
+             * authenticate and verify credentials.
+             *
+             * In subsequent authentication flows where a TGT is available and only an ST needs to be
+             * created, there are no cached copies of the credential, since we do have a TGT available.
+             * So we will simply grab the available authentication and produce the final result based on that.
+             */
+            final Authentication authentication = ticketRegistrySupport.getAuthenticationFrom(ticketGrantingTicket);
+            if (authentication == null) {
+                throw new InvalidTicketException(
+                        new AuthenticationException("No authentication found for ticket " + ticketGrantingTicket), ticketGrantingTicket);
+            }
+            final AuthenticationResultBuilder authenticationResultBuilder = authenticationSystemSupport
```

Comment:
OK, roger that. +1 to merge

On Thu, Feb 4, 2016 at 07:38 Misagh Moayyed <notifications@github.com>
wrote:

> In
> cas-server-webapp-actions/src/main/java/org/jasig/cas/web/flow/GenerateServiceTicketAction.java
> <https://github.com/Jasig/cas/pull/1511#discussion_r51866805>:
>
> > -                    this.authenticationSystemSupport.handleAndFinalizeSingleAuthenticationTransaction(service, credential);
> > +            /**
> > +             * In the initial primary authentication flow, credentials are cached and available.
> > +             * Since they are authenticated as part of submission first, there is no need to doubly
> > +             * authenticate and verify credentials.
> > +             *
> > +             * In subsequent authentication flows where a TGT is available and only an ST needs to be
> > +             * created, there are no cached copies of the credential, since we do have a TGT available.
> > +             * So we will simply grab the available authentication and produce the final result based on that.
> > +             */
> > +            final Authentication authentication = ticketRegistrySupport.getAuthenticationFrom(ticketGrantingTicket);
> > +            if (authentication == null) {
> > +                throw new InvalidTicketException(
> > +                        new AuthenticationException("No authentication found for ticket " + ticketGrantingTicket), ticketGrantingTicket);
> > +            }
> > +            final AuthenticationResultBuilder authenticationResultBuilder = authenticationSystemSupport
>
> The API change you see above is in fact borrowed from the MFA branch, just
> so I could keep it all consistent and have it merge easier later.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Jasig/cas/pull/1511/files#r51866805>.
>